source/dotnet/Library/AdaptiveCards/ParseContext.cs

```diff
@@ -0,0 +1,131 @@
+        public static void PopElement()
+        {
+            // about to pop an element off the stack. perform collision list maintenance and detection.
+            var idsToPop = IDStack.Peek();
+            string elementID = idsToPop.Item1;
+            AdaptiveInternalID elementInternalID = idsToPop.Item2;
+            bool isFallback = idsToPop.Item3;
+
+            if (!string.IsNullOrEmpty(elementID))
+            {
+                bool haveCollision = false;
+                var nearestFallbackID = GetNearestFallbackID(elementInternalID);
+
+                // Walk through the list of elements we've seen with this ID
+                if (ElementIDs.ContainsKey(elementID))
+                {
+                    foreach (var entryFallBackID in ElementIDs[elementID])
+                    {
+                        // If the element we're about to pop is the fallback parent for this entry, then there's no collision
+                        // (fallback content is allowed to have the same ID as its parent)
+                        if (entryFallBackID.Equals(elementInternalID))
+                        {
+                            haveCollision = false;
+                            break;
+                        }
+
+                        // The inverse of the above -- if this element's fallback parent is the entry we're looking at, there's
+                        // no collision.
+                        try
+                        {
+                            // 0 is the last item on the stack (the one we're about to pop)
+                            // 1 is the parent of the last item on the stack
+                            var previousInStack = IDStack.ElementAt(1);
+
+                            if (previousInStack.Item2.Equals(entryFallBackID))
+                            {
+                                // we're looking at a fallback entry for our parent
+                                break;
+                            }
+                        }
+                        catch (ArgumentOutOfRangeException)
+                        {
+                            // we're looking at a toplevel element
+                        }
+
+                        // if the element we're inspecting is fallback content, continue on to the next entry
+                        if (isFallback)
+                        {
+                            continue;
+                        }
+
+                        // at this point, we may or may not have a collision depending on additional entries.
+                        haveCollision = true;
+                    }
+                }
+
+                if (haveCollision)
+                {
+                    throw new AdaptiveSerializationException("Collision detected for id '" + elementID + "'");
+                }
+
+                // no need to add an entry for this element if it's fallback (we'll add one when we parse it for non-fallback)
```

Comment:
Nit: Maybe have teh comment match the code it sits adjacent to :).. comment talks of what shouldnt happen in the fallback case but the code really is handling what happens in the non-fallback case.